---
title: "Single Gene Report V1.4"
author: "Yilin Bai"
date: "2023-09-19"
output: 
  bookdown::html_document2:
    code_folding: show
    number_sections: no
    toc: yes
    toc_depth: 6
    toc_float: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,message = FALSE, warning = FALSE)
```

## Reference

**Dataset:** Suvorov, Anton; Kim, Bernard; Wang, Jeremy; Armstrong, Ellie; Peede, David; D'Agostino, Emmanuel R. R.; et al. (2020). Widespread introgression across a phylogeny of 155 Drosophila genomes. figshare. Dataset. <https://doi.org/10.6084/m9.figshare.13264697.v1>

**Paper:** Suvorov A, Kim BY, Wang J, Armstrong EE, Peede D, D'Agostino ERR, Price DK, Waddell P, Lang M, Courtier-Orgogozo V, David JR, Petrov D, Matute DR, Schrider DR, Comeault AA. Widespread introgression across a phylogeny of 155 Drosophila genomes. Curr Biol. 2022 Jan 10;32(1):111-123.e5. doi: 10.1016/j.cub.2021.10.052. Epub 2021 Nov 16. PMID: 34788634; PMCID: PMC8752469.

## Target alignment

**Outgroup:** Anopheles

**Type:** Aligned DNA BUSCO loci

## Run the command

## Alignment summary

```{r}
library(Biostrings)

candidate <- 1
# "current_dir" is the folder where you save a set of multi-alignments
current_dir <- getwd()
gene_sets <- system2(command = "ls", args = current_dir, stdout = TRUE)
cat("Current gene file location:", current_dir, "\n")
gene_file <- gene_sets[candidate]
gene_name <- sub("\\..*", "", gene_file)

store_path <- paste(current_dir,'/',gene_name,'/', sep = "")

# Set the path and filename for iqtree running result
prefix_single <- paste(current_dir,"/","Single_", gene_name, sep = "")
prefix_mix <- paste(current_dir,"/","Mix_", gene_name, sep = "")
# Set the command for both single and mix model
arg_single <- c("-s", gene_file, "-B", "1000", "--prefix", prefix_single)
arg_mix <- c("-s", gene_file, "-m", "ESTMIXNUM", "-mrate", "E,I,G,I+G,R,I+R", "-opt_qmix_criteria", "1", "--prefix", prefix_mix)

# The species tree file should be assigned manually
outgroup <- "Anopheles_gambiae"
cat("Current gene name:", gene_sets[candidate], "\n")
gene_alignment <- readDNAStringSet(gene_file)
file_name <- basename(as.character(gene_alignment))


# Command of run one class model iq-tree
system2(command = "/Users/lindsaybai/Desktop/BIOL8706/iqtree-2.2.2.7.modelmix-MacOSX/bin/iqtree2", args = arg_single, stdout = FALSE)
# Command of run mixture class model iq-tree
system2(command = "/Users/lindsaybai/Desktop/BIOL8706/iqtree-2.2.2.7.modelmix-MacOSX/bin/iqtree2", args = arg_mix, stdout = FALSE)

```

```{r}
# Read the tree 
single_iqtree <- list.files(current_dir, pattern = "^Single.*\\.iqtree$", full.names = TRUE)
mixture_iqtree <- list.files(current_dir, pattern = "^Mix.*\\.iqtree$", full.names = TRUE)

one_line <- readLines(single_iqtree)
mix_line <- readLines(mixture_iqtree)
```

```{r}
# length and Ntax
Length_line <- grep("Input data: ", one_line, value = TRUE)
Length_matches <- gregexpr("\\d+", Length_line)
Length_numbers <- regmatches(Length_line, Length_matches)[[1]]
# Convert the extracted strings to integers
Ntaxa <- as.integer(Length_numbers[1])
Length <- as.integer(Length_numbers[2])
```

```{r}
# percent_constant & percent_parsimony_informative
percent_constant_line <- grep("Number of constant sites: ", one_line, value = TRUE)
percent_constant <- sprintf("%.5s%%", regmatches(percent_constant_line, regexpr("\\d+\\.\\d+%\\s+", percent_constant_line)))
parsimony_informative <- as.numeric(sub("Number of parsimony informative sites: ","\\1",grep("Number of parsimony informative sites: ", one_line, value = TRUE))) 
percent_parsimony_informative <- sprintf("%.2f%%", parsimony_informative / Length * 100)
site_patterns <- as.numeric(sub("Number of distinct site patterns: ","\\1",grep("Number of distinct site patterns: ", one_line, value = TRUE)))
```

```{r}
# Create the data frame
alignment_data <- data.frame(
  "Locus_name" = gene_name,
  "Length" = Length,
  "Ntaxa" = Ntaxa,
  "Percent_constant" = percent_constant,
  "Percent_parsimony_informative" = percent_parsimony_informative,
  "Site_patterns" = site_patterns
)
alignment_data
```

## Tree Topology

```{r}
library(ape)
library(phytools)
# Read the tree files
Mixture_tree = read.tree(list.files(current_dir, pattern = "^Single.*\\.treefile$", full.names = TRUE))
One_class_tree = read.tree(list.files(current_dir, pattern = "^Mix.*\\.treefile$", full.names = TRUE))
Species_tree = read.tree("/Users/lindsaybai/Desktop/BIOL8706/phylogenetic_project/drosophila_introgression_data/tree/astral.tre")
```

### Phylograms plot


```{r}
library("ggtree")
library("ggplot2")

cladograms <- function(tree, y_scale = -5, subtitle = "") {
  ggtree(tree, cex = 0.8, layout = "slanted", branch.length = "none") +
    geom_tiplab(align = TRUE, size = 2) +
    geom_treescale(y = y_scale, color = "black", fontsize = 4) +
    labs(subtitle = subtitle) + 
    theme(legend.position = "bottom")
}

p1 <- cladograms(One_class_tree, subtitle = "Single Model")
p2 <- cladograms(Mixture_tree, subtitle = "Mix Model")

multiplot(p1, p2, ncol = 2)

phylograms <- function(tree, y_scale = -5, subtitle = "") {
  ggtree(tree, cex = 0.8, aes(color = branch.length - mean(branch.length))) +
    scale_color_continuous(
      high = "red", low = "black",
      name = "centralized(Branch Length)",
      guide = guide_colourbar(barwidth = 10)
    ) +
    geom_tiplab(align = TRUE, size = 2) +
    geom_treescale(y = y_scale, color = "black", fontsize = 4) +
    labs(subtitle = subtitle) +
    theme(legend.position = "bottom")
}

p1 <- phylograms(One_class_tree, subtitle = "Single Model")
p2 <- phylograms(Mixture_tree, subtitle = "Mix Model")

multiplot(p1, p2, ncol = 2)
```

### Cophylogeny plot

```{r}
## create co-phylogenetic object
wasp.cophylo<-cophylo(Mixture_tree,Mixture_tree)
## plot co-phylogenies
plot(wasp.cophylo,link.type="curved",link.lwd=4,
 link.lty="solid",link.col=make.transparent("red",
 0.25))
par(mar=c(5.1,4.1,4.1,2.1))
```


```{r}
library(ape)
library(phangorn)

taxa_single <- One_class_tree$tip.label
taxa_mix <- Mixture_tree$tip.label
# Check if there are any different elements between taxa_single and taxa_mix
taxa_equal_model <- setequal(taxa_single, taxa_mix)
if (!taxa_equal_model){
  taxa_diff <- union(setdiff(taxa_single, taxa_mix),setdiff(taxa_mix, taxa_single))
  tree_single <- drop.tip(tree_single, taxa_diff)
  tree_mix <- drop.tip(tree_mix, taxa_diff)
  cat("The taxa of the two model trees are not equal, the taxa of the two trees are unified.\n")
}


# Find the intersection of taxa
taxa_intersect <- intersect(One_class_tree$tip.label, Mixture_tree$tip.label)
taxa_subset <- intersect(taxa_intersect, Species_tree$tip.label)

# Trim the species tree
tree_species_trimed <- drop.tip(Species_tree, setdiff(Species_tree$tip.label, taxa_subset))


# Placeholder data (replace these with your actual data)
tree_names <- c("One_class", "Mixture", "Species")
tree_files <- list(
  One_class = One_class_tree,
  Mixture = Mixture_tree,
  Species = tree_species_trimed
)


RF_dists <- list()
nRF_dists <- list()
wRF_dists <- list()
KF_dists <- list()
PD_dists <- list()
wPD_dists <- list()
```

```{r}
# Loop through the combinations of gene names and tree pairs
tree_combinations <- combn(length(tree_files), 2, simplify = FALSE)
  for (comb in tree_combinations) {
      i <- comb[1]
      j <- comb[2]
        
      tree1 <- tree_files[[i]]
      tree2 <- tree_files[[j]]
      tree1_name <- tree_names[i]
      tree2_name <- tree_names[j]
      
      # Placeholder for distance calculation (replace with actual distance calculation function)
      RF_dist <- RF.dist(tree1, tree2, normalize = FALSE, check.labels = TRUE, rooted = FALSE)
      nRF_dist <-RF.dist(tree1, tree2, normalize = TRUE, check.labels = TRUE, rooted = FALSE)
      wRF_dist <- wRF.dist(tree1, tree2, normalize = FALSE, check.labels = TRUE, rooted = FALSE)
      KF_dist <- KF.dist(tree1, tree2, check.labels = TRUE, rooted = FALSE)
      PD_dist <- path.dist(tree1, tree2, check.labels = TRUE, use.weight = FALSE)
      wPD_dist <- path.dist(tree1, tree2, check.labels = TRUE, use.weight = TRUE)
      
      RF_dists <- append((RF_dists), list(RF_dist))
      nRF_dists <- append(nRF_dists, list(nRF_dist))
      wRF_dists <- append(wRF_dists, list(wRF_dist))
      KF_dists <- append(KF_dists, list(KF_dist))
      PD_dists <- append(PD_dists, list(PD_dist))
      wPD_dists <- append(wPD_dists, list(wPD_dist))

      
  }

RF_dists <- unlist(RF_dists)
nRF_dists <- unlist(nRF_dists)
wRF_dists <- unlist(wRF_dists)
KF_dists <- unlist(KF_dists)
PD_dists <- unlist(PD_dists)
wPD_dists <- unlist(wPD_dists)


# Add the results to the data frame


distance_table <- data.frame(
        RF_vs_gene_tree = rep(RF_dists[1],2),
        RF_vs_species_tree = c(RF_dists[2],RF_dists[3]),
        nRF_vs_gene_tree = rep(nRF_dists[1],2),
        nRF_vs_species_tree = c(nRF_dists[2],nRF_dists[3]),
        wRF_vs_gene_tree = rep(wRF_dists[1],2),
        wRF_vs_species_tree = c(wRF_dists[2],wRF_dists[3]),        
        KF_vs_gene_tree = rep(KF_dists[1],2),
        KF_vs_species_tree = c(PD_dists[2],PD_dists[3]),
        PD_vs_gene_tree = rep(PD_dists[1],2),
        PD_vs_species_tree = c(PD_dists[2],PD_dists[3]),
        wPD_vs_gene_tree = rep(wPD_dists[1],2),
        wPD_vs_species_tree = c(wPD_dists[2],wPD_dists[3]),
        row.names = c("One class model","Mixture class model"))

```

## Branch Lengths

### Summary parameters table

```{r}
# tree length & Sum_int & prop_int
one_length <- sub("Total tree length \\(sum of branch lengths\\): ","\\1",grep("Total tree length \\(sum of branch lengths\\): ", one_line, value = TRUE))
mix_length <- sub("Total tree length \\(sum of branch lengths\\): ","\\1",grep("Total tree length \\(sum of branch lengths\\): ", mix_line, value = TRUE))

one_sum_int <- sub("Sum of internal branch lengths: ([0-9.]+).*","\\1",grep("Sum of internal branch lengths: ([0-9.]+).*", one_line, value = TRUE))
mix_sum_int <- sub("Sum of internal branch lengths: ([0-9.]+).*","\\1",grep("Sum of internal branch lengths: ([0-9.]+).*", mix_line, value = TRUE))

one_prop_int <- sub(" of tree length", "", sub(".*\\(([^)]+)\\).*","\\1",grep("Sum of internal branch lengths: ([0-9.]+).*", one_line, value = TRUE)))
mix_prop_int <- sub(" of tree length", "", sub(".*\\(([^)]+)\\).*","\\1",grep("Sum of internal branch lengths: ([0-9.]+).*", mix_line, value = TRUE)))
```

```{r}
# Provided Newick tree string
Mixture_txt <- readLines(list.files(current_dir, pattern = "^Mix.*\\.treefile$"))
One_class_txt <- readLines(list.files(current_dir, pattern = "^Single.*\\.treefile$"))

branch_len_mix <- as.numeric(gsub("([0-9.]+).*", "\\1", strsplit(Mixture_txt, ":")[[1]])[-1])
branch_len_one <- as.numeric(gsub("([0-9.]+).*", "\\1", strsplit(One_class_txt, ":")[[1]])[-1])

sum_branch_len_one <- summary(branch_len_one)
sum_branch_len_mix <- summary(branch_len_mix)
```

```{r}
# Create a dataframe with the provided summary statistics
df <- data.frame(
  Tree_Length = c(one_length, mix_length),
  Sum_int = c(one_sum_int, mix_sum_int),
  prop_int = c(one_prop_int, mix_prop_int),
  min = c(sum_branch_len_one[1], sum_branch_len_mix[1]),
  Qu_1st = c(sum_branch_len_one[2], sum_branch_len_mix[2]),
  Median = c(sum_branch_len_one[3], sum_branch_len_mix[3]),
  Mean = c(sum_branch_len_one[4], sum_branch_len_mix[4]),
  Qu_3rd = c(sum_branch_len_one[5], sum_branch_len_mix[5]),
  Max = c(sum_branch_len_one[6], sum_branch_len_mix[6]),
  row.names = c("One model", "Mixture model"))
# Print the dataframe
print(df)
```

### Faceted histogram

```{r}
# Assuming you have loaded the 'ggplot2' package and have the necessary data

data <- data.frame(
  model = rep(c("One class", "Mixture"), each = length(branch_len_one)),
  branch_length = c(branch_len_one, branch_len_mix)
)

# Create a faceted histogram
library("ggplot2")
ggplot(data, aes(x = branch_length)) +
  geom_histogram(binwidth = 0.1, fill = "blue", color = "black") +
  facet_grid(model ~ ., scales = "free_y") +  # Facet by 'model', free y-axis scales
  labs(x = "Branch Length", y = "Frequency") +
  theme_minimal()+
  scale_x_log10()
```

### ECDF plot

```{r}
# Create an ECDF plot
ggplot(data, aes(x = branch_length, color = model)) +
  stat_ecdf(geom = "step") +
  labs(x = "Branch Length", y = "ECDF") +
  scale_color_manual(values = c("One class" = "blue", "Mixture" = "red")) +
  theme_minimal() +
  theme(legend.position = "top")+
  scale_x_log10()
```

## Models

### Description

1. One Class Model

```{r}
library(data.table)
library(stringr)
one_model <- grep("^Model of substitution:", one_line, value = TRUE)
one_class_mod_lines <- one_line[(which(one_line == one_model)):(which(one_line == one_model)+9)]
Model_of_substitution <- sub("^Model of substitution: ","\\1",one_model)
RP_AC <- sub("A-C: ","\\1",grep("A-C: ", one_class_mod_lines[5], value = TRUE))
RP_AG <- sub("A-G: ","\\1",grep("A-G: ", one_class_mod_lines[6], value = TRUE))
RP_AT <- sub("A-T: ","\\1",grep("A-T: ", one_class_mod_lines[7], value = TRUE))
RP_CG <- sub("C-G: ","\\1",grep("C-G: ", one_class_mod_lines[8], value = TRUE))
RP_CT <- sub("C-T: ","\\1",grep("C-T: ", one_class_mod_lines[9], value = TRUE))
RP_GT <- sub("G-T: ","\\1",grep("G-T: ", one_class_mod_lines[10], value = TRUE))

one_class_mod_table <- data.frame("Model_of_substitution" = Model_of_substitution,
                                  "Rate_parameter_AC" = RP_AC,
                                  "Rate_parameter_AG" = RP_AG,
                                  "Rate_parameter_AT" = RP_AT,
                                  "Rate_parameter_CG" = RP_CG,
                                  "Rate_parameter_CT" = RP_CT,
                                  "Rate_parameter_GT" = RP_GT,
                                  row.names = "One class model")
```

2. Mix Class Model

```{r}
mix_model <- grep("^Mixture model of substitution:", mix_line, value = TRUE)
mix_df1 <- data.table(Line = mix_model)
setnames(mix_df1, new = "Mix Class Model")
```

```{r}
# Example lines
Component_mix_model <- mix_line[(which(mix_line == mix_model)+3):(which(mix_line == mix_model)+6)]
# Initialize an empty vector to store words
all_words <- c()
# Iterate through each line
for (line in Component_mix_model) {
words <- strsplit(line, " ")[[1]]
      non_empty_words <- words[words != ""]
      all_words <- c(all_words, non_empty_words)
    }

    matrix_data <- matrix(all_words,
                       nrow = 3, ncol = 5, byrow = TRUE)

mixture_model_of_substitution <- sub("^Mixture model of substitution: ","\\1",mix_model)

mix_class_mod_table <- data.frame("Model_of_substitution" = mixture_model_of_substitution,
                                  "Component_1" = all_words[2],
                                  "Rate_1" = all_words[3],
                                  "Weight_1" = all_words[4],
                                  "Parameters_1" = all_words[5],
                                  "Component_2" = all_words[7],
                                  "Rate_2" = all_words[8],
                                  "Weight_2" = all_words[9],
                                  "Parameters_2" = all_words[10],
                                  "Component_3" = all_words[12],
                                  "Rate_3" = all_words[13],
                                  "Weight_3" = all_words[14],
                                  "Parameters_3" = all_words[15],
                                  row.names = "Mixture class model")
mix_class_mod_table

```

```{r}
library(tidyverse)
Model_of_substitution_table <- bind_rows(one_class_mod_table,mix_class_mod_table)
```

### Summary table

```{r}
one_model <- sub("^Model of substitution: ","\\1",grep("^Model of substitution:", one_line, value = TRUE))
mix_model <- sub("^Mixture model of substitution:", "\\1", mix_model <- grep("^Mixture model of substitution:", mix_line, value = TRUE))
```

```{r}
# Rates, Likelihood, Unconstrained_likelihood
one_rates <- sub("Model of rate heterogeneity: ","\\1",grep("Model of rate heterogeneity: ", one_line, value = TRUE)) 
mix_rates <-sub("Model of rate heterogeneity: ","\\1",grep("Model of rate heterogeneity: ", mix_line, value = TRUE)) 
one_likelihood <- sub("Log-likelihood of the tree: " ,"\\1",grep("Log-likelihood of the tree: ", one_line, value = TRUE)) 
mix_likelihood <- sub("Log-likelihood of the tree: " ,"\\1",grep("Log-likelihood of the tree: ", mix_line, value = TRUE)) 

one_uncons_likelihood <- sub("Unconstrained log-likelihood \\(without tree\\):","\\1",grep("Unconstrained log-likelihood \\(without tree\\):", one_line, value = TRUE)) 
mix_uncons_likelihood <- sub("Unconstrained log-likelihood \\(without tree\\):","\\1",grep("Unconstrained log-likelihood \\(without tree\\):", mix_line, value = TRUE)) 
```

```{r}
# parameters
one_para <- sub("Number of free parameters \\(#branches \\+ #model parameters\\): ","\\1",grep("Number of free parameters \\(#branches \\+ #model parameters\\): ", one_line, value = TRUE))
mix_para <- sub("Number of free parameters \\(#branches \\+ #model parameters\\): ","\\1",grep("Number of free parameters \\(#branches \\+ #model parameters\\): ", mix_line, value = TRUE))
```

```{r}
# AIC,AICc, BIC,
one_AIC <- sub("Akaike information criterion \\(AIC\\) score: ","\\1",grep("Akaike information criterion \\(AIC\\) score: ", one_line, value = TRUE))
mix_AIC <- sub("Akaike information criterion \\(AIC\\) score: ","\\1",grep("Akaike information criterion \\(AIC\\) score: ", mix_line, value = TRUE))

one_AICC <- sub("Corrected Akaike information criterion \\(AICc\\) score: ","\\1",grep("Corrected Akaike information criterion \\(AICc\\) score: ", one_line, value = TRUE))
mix_AICC <- sub("Corrected Akaike information criterion \\(AICc\\) score: ","\\1",grep("Corrected Akaike information criterion \\(AICc\\) score: ", mix_line, value = TRUE))

one_BIC <- sub("Bayesian information criterion \\(BIC\\) score: ","\\1",grep("Bayesian information criterion \\(BIC\\) score: ", one_line, value = TRUE))
mix_BIC <- sub("Bayesian information criterion \\(BIC\\) score: ","\\1",grep("Bayesian information criterion \\(BIC\\) score: ", mix_line, value = TRUE))

if (one_BIC < mix_BIC) {
  best_model <- "One"
} else if (mix_BIC < one_BIC) {
  best_model <- "Mixture"
} else {
  best_model <- "Both models have the same BIC score"
}
```

```{r}
# Create the data frame
model_data <- data.frame(
  row.names = c("One model", "Mixture model"),
  Classes = c(1, 3),
  Model_String = c(one_model, mix_model),
  Rates = c(one_rates, mix_rates),
  Likelihood = c(one_likelihood,mix_likelihood),
  Unconstrained_likelihood = c(one_uncons_likelihood,mix_uncons_likelihood),
  parameters = c(one_para,mix_para),
  AIC = c(one_AIC,mix_AIC),
  AICc = c(one_AICC,mix_AICC),
  BIC = c(one_BIC, mix_BIC),
  Best = rep(best_model,2)
)
```


```{r}
combined_table <- bind_cols(alignment_data,Model_of_substitution_table,df,distance_table,model_data)

file_name <- paste(gene_name,'.csv', sep = "")
write.csv(combined_table, file = file_name, row.names = FALSE)
```
